src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs

```diff
@@ -0,0 +1,6195 @@
+        {
+            var src = @"
+struct C
+{
+    public Buffer10<Buffer10<int>> F;
+
+    [System.Diagnostics.CodeAnalysis.UnscopedRef]
+    public ref int M2()
+    {
+        ref int y = ref F[0][0];
+        return ref y;
+    }
+}
+
+class Program
+{
+    static void Main()
+    {
+        var x = new C();
+        x.M2() = 111;
+        System.Console.Write(x.F[0][0]);
+    }
+}
+";
+            var comp = CreateCompilation(src + Buffer10Definition, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseExe);
+            var verifier = CompileAndVerify(comp, expectedOutput: "111", verify: Verification.Fails).VerifyDiagnostics();
+        }
+
+        [ConditionalFact(typeof(MonoOrCoreClrOnly))]
+        public void ElementAccess_Index_Variable_01()
+        {
+            var src = @"
+class C
+{
+    public Buffer10<int> F;
+}
+
+class Program
+{
+    static void Main()
+    {
+        var x = new C();
+        System.Console.Write(M1(x));
+        M2(x);
+        System.Console.Write(' ');
+        System.Console.Write(M1(x));
+    }
+
+    static int M1(C x) => x.F[^10];
+    static void M2(C x) => x.F[^10] = 111;
+}
+";
+            var comp = CreateCompilation(src + Buffer10Definition, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseExe);
+            var verifier = CompileAndVerify(comp, expectedOutput: "0 111", verify: Verification.Fails).VerifyDiagnostics();
+
+            verifier.VerifyIL("Program.M1",
+@"
+{
+  // Code size       24 (0x18)
+  .maxstack  2
+  .locals init (System.Span<int> V_0)
+  IL_0000:  ldarg.0
+  IL_0001:  ldflda     ""Buffer10<int> C.F""
+  IL_0006:  ldc.i4.s   10
+  IL_0008:  call       ""InlineArrayAsSpan<Buffer10<int>, int>(ref Buffer10<int>, int)""
+  IL_000d:  stloc.0
+  IL_000e:  ldloca.s   V_0
+  IL_0010:  ldc.i4.0
+  IL_0011:  call       ""ref int System.Span<int>.this[int].get""
+  IL_0016:  ldind.i4
+  IL_0017:  ret
+}
+");
+
+            verifier.VerifyIL("Program.M2",
+@"
+{
+  // Code size       26 (0x1a)
+  .maxstack  2
+  .locals init (System.Span<int> V_0)
+  IL_0000:  ldarg.0
+  IL_0001:  ldflda     ""Buffer10<int> C.F""
+  IL_0006:  ldc.i4.s   10
+  IL_0008:  call       ""InlineArrayAsSpan<Buffer10<int>, int>(ref Buffer10<int>, int)""
+  IL_000d:  stloc.0
+  IL_000e:  ldloca.s   V_0
+  IL_0010:  ldc.i4.0
+  IL_0011:  call       ""ref int System.Span<int>.this[int].get""
+  IL_0016:  ldc.i4.s   111
+  IL_0018:  stind.i4
+  IL_0019:  ret
+}
+");
+        }
+
+        [ConditionalFact(typeof(MonoOrCoreClrOnly))]
+        public void ElementAccess_Dynamic_Variable_01()
+        {
+            var src = @"
+class C
+{
+    public Buffer10<int> F;
+}
+
+class Program
+{
+    static void Main()
+    {
+        var x = new C();
+        System.Console.Write(M1(x));
+        M2(x);
+        System.Console.Write(' ');
+        System.Console.Write(M1(x));
+    }
+
+    static int M1(C x) => x.F[0];
+    static void M2(C x) => x.F[(dynamic)0] = 111;
```

Comment:
Consider adding a test indexing with a `dynamic` variable, not constant. Also, do we have tests indexing with `int` variables (not constants)?

Reply:
> Consider adding a test indexing with a dynamic variable, not constant. 

There is no such thing a s a dynamic constant, unless it has `null` value. Which is not the case here.

> Also, do we have tests indexing with int variables (not constants)?

I do not find this scenario particularly interesting. But will include it into the test plan for completeness.